File: April/Abiji-2020/C++/Open the Lock /Open the lock.cpp
class Solution {
public:
    int openLock(vector<string>& deadends, string target) {
        int pow10[] = {1,10,100,1000};
        int visit[10000]={0};
        for(string dead:deadends){
            visit[stoi(dead)] = 2;
        }
        int src =0, dest = stoi(target), steps =0 , dir =1;
        if(visit[src]==2 || visit[dest]==2) return -1;
        if (src == dest) return 0;
        queue<int> forward, backward;
        forward.push(src);
        visit[src] = 1;
        backward.push(dest);
        visit[dest] = -1;
        while(!forward.empty() && !backward.empty()){
            if (forward.size()> backward.size()){
                swap(forward,backward);
                dir = -dir;
            }
            steps++;
            int size = forward.size();
            while(size-- > 0){
                int curr = forward.front();
                forward.pop();
                for(int p:pow10){
                    int d = (curr/p)%10;
                    for(int i=-1;i<=1;i+=2){
                        int z = d+i;
                        z= z==-1 ? 9 : (z==10 ? 0:z);
                        int next = curr+(z-d)*p;
                        if(visit[next]==-dir) return steps;
                        if(visit[next]==0){
                            forward.push(next);
                            visit[next] = dir;
                        }
                    }
                }
            }
        }
        return -1;
   }
};

File: April/Abiji-2020/Golang /Open the Lock/open the lock.go

func openLock(deadends []string, target string) int {
	pow10 := []int{1, 10, 100, 1000}
	visit := make(map[int]int)

	for _, dead := range deadends {
		deadNum, _ := strconv.Atoi(dead)
		visit[deadNum] = 2
	}

	dest, _ := strconv.Atoi(target)
	if visit[0] == 2 || visit[dest] == 2 {
		return -1
	}
	if 0 == dest {
		return 0
	}

	src, steps, dir := 0, 0, 1
	forward, backward := make([]int, 0), make([]int, 0)
	forward = append(forward, src)
	visit[src] = 1
	backward = append(backward, dest)
	visit[dest] = -1

	for len(forward) > 0 && len(backward) > 0 {
		if len(forward) > len(backward) {
			forward, backward = backward, forward
			dir = -dir
		}
		steps++
		size := len(forward)
		for size > 0 {
			curr := forward[0]
			forward = forward[1:]
			size--
			for _, p := range pow10 {
				d := (curr / p) % 10
				for _, i := range []int{-1, 1} {
					z := d + i
					if z == -1 {
						z = 9
					} else if z == 10 {
						z = 0
					}
					next := curr + (z-d)*p
					if visit[next] == -dir {
						return steps
					}
					if visit[next] == 0 {
						forward = append(forward, next)
						visit[next] = dir
					}
				}
			}
		}
	}
	return -1
}
